feat: Multiple Private links per Event Type

Souptik2001 opened the pull request:
## What does this PR do?

Implements single-use links feature!

Disclaimer: Contains schema changes and migrations. ⚠️ 

- Fixes #15776 
- Fixes #16228

https://www.loom.com/share/cb93616c775a4704bdc7590c6c1aa58e?sid=06bb210e-fd7b-4772-827a-e1d77e460ff3

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Please check the above loom video, which contains the detailed walkthrough. 👆

Souptik2001 committed:
feat: Change hashedLink field to array

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
feat: Implement core single-use links logic

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Get private link from list of hashedLinks on event list page

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix defalut single-use links form field value

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix some type errors while compiling

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Use onChange callback instead of setting form values directly

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -553,6 +558,8 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
     const updatedFields: Partial<FormValues> = {};
     Object.keys(dirtyFields).forEach((key) => {
       const typedKey = key as keyof typeof dirtyFields;
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
       updatedFields[typedKey] = undefined;
```
For some reason getting this type error on this line, and at all not able to fix this, therefore added this. If you have some solution please let me know. 🙏 
FYI - few lines below if you see [for this line](https://github.com/calcom/cal.com/pull/15896/files#diff-df6d4a41111a432e757de2e1297681f80d9a8d9d74641c89e8f1a12333d4a09aR568) the same `ts-ignore` is added, maybe it was also getting the same error.

`Type error: Expression produces a union type that is too complex to represent.`

Souptik2001 commented:
Also one weird thing which is happening is that, for the newly added texts, like the "single_use_link_title", etc., when we change the tab the strings get's changed to the placeholder key, as you can see in the video.

Not sure if I am missing something related to this. 🤔

anikdhabal reviewed (commented):
Great work @Souptik2001 

I've noticed in the Loom video that after destroying a single-use link, the page shows a 404 error, which is expected. However, it doesn't look good. Could you manage our own custom 404 page instead?

anikdhabal commented:
> Also one weird thing which is happening is that, for the newly added texts, like the "single_use_link_title", etc., when we change the tab the strings get's changed to the placeholder key, as you can see in the video.
> 
> Not sure if I am missing something related to this. 🤔

Every time you add a new translation key to the locales, you need to restart the server in order to see the text string

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
Hey, please make this a setting toggle component, just like the other options. When enabled, it should create one link at that time. If users want more links, they can add them using the 'Add New Link' button, similar to how it currently works.

anikdhabal reviewed (commented):
Hey, please make this a setting toggle component, just like the other options. When enabled, it should create one link at that time. If users want more links, they can add them using the 'Add New Link' button, similar to how it currently works.

Souptik2001 commented:
@anikdhabal 

> I've noticed in the Loom video that after destroying a single-use link, the page shows a 404 error, which is expected. However, it doesn't look good. Could you manage our own custom 404 page instead?

Actually it is using the default 404 page only. I have not created my own 404 page. As per my best guess, it is happening due to this PR - https://github.com/calcom/cal.com/pull/15696
Please correct me if I am wrong. 🙏 

> Every time you add a new translation key to the locales, you need to restart the server in order to see the text string

I have actually restarted the server. Infact when I was recording the video, I was even not in dev mode. And as I said it is initially appearing fine (correct text is coming), but as soon as I change the tab and come back to the tab it is somehow changing to the translation key. I tried so much, but couldn't find a way to solve it. 😅

anikdhabal reviewed (changes requested):
After watching the Loom video, I feel some changes are needed. Please address it first

Souptik2001 commented:
> Hey, please make this a setting toggle component, just like the other options. When enabled, it should create one link at that time. If users want more links, they can add them using the 'Add New Link' button, similar to how it currently works.

@anikdhabal Before making this major change, just wanted to confirm few things. Please let me know what do you think about this..

I initially thought about making it a toggle, but then thought that this was more like an "adding feature" than a "toggling feature".

If we add the toggle feature, then certain complexities/questions popup like -
- What will happen when the user deletes the last remaining link? Will the toggle automatically close?
- Will this "adding" behavior not make it little bit confusing with the private link feature, as the user will not be able to see the  "add new link" button in the first place.

Considering these only I thought to simply make it like this.

Just taking suggestions. Please let me know your thoughts on this. 🙌

anikdhabal commented:
@Souptik2001 take this as an example:-

https://github.com/user-attachments/assets/c28f2ecb-f3e7-4b9c-825c-a4cc0df4a196

And when user deletes the last remaining link, the "add new link" button visible in the first place. How is now looking
![Screenshot 2024-07-25 014538](https://github.com/user-attachments/assets/73cff231-589a-4e2d-a332-8925b1ede228)

Currently, the description is not structured; it needs to be more descriptive.

This is my thought on the improvements, but I suggest waiting until someone from the design approves your implementation and design.

Tagging @ciaranha to check this once.

Souptik2001 commented:
Thanks @anikdhabal got it!

Ok waiting for @ciaranha's confirmation before proceeding on your suggested changes!

Souptik2001 commented:
Hey @ciaranha Any update on this? 🙌

keithwillcode reviewed (dismissed):
Thank you for your contribution. 

to be merged, the logical changes in the backend need unit/integration tests to be added.

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 committed:
feat: Add e2e tests for single-use links

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Thanks @keithwillcode . Added e2e tests covering all scenarios of single-use links -

- Single use links can be added.
- Single use links can be removed.
- Single use links should be auto-destroyed on use.

Please let me know if anything else needs to be added.

Here's a video of the last test - 

https://www.loom.com/share/a24e5fa1ba5f4469b01af12429e914b8?sid=1b898285-d84d-44c4-b867-3d6aca7685f0

cc: @anikdhabal

anikdhabal commented:
Hey @Souptik2001, I think you might not fully understand @keithwillcode’s suggestion. You need to add unit and integration tests for the backend logical changes. Additionally, we need tests for the frontend logical part as well for this PR. However, the frontend part still needs approval, so please add the tests for it after everything is properly set in the frontend.

We always avoid creating new end-to-end (e2e) tests as they tend to be flaky and slow. We prefer unit testing instead. So, if possible, please add unit tests for this feature.

Thanks!

Souptik2001 commented:
Ohhh got it @anikdhabal 

One request - Can you please give me example of a test which quite some examples of testing different types of things. 🙏 
The playwright tests are arranged properly therefore understood those. But having some difficultly in understanding the unit tests of this project..

> We always avoid creating new end-to-end (e2e) tests as they tend to be flaky and slow.

💯% agree. I had to struggle so much while writing these three tests 😅

CarinaWolli commented:
Can you make sure that the design looks like the other settings? With the toggle to expand the settings? Also, font size is different and the subtitle should be right under the title with a line before the expanded settings start. 

We have the `SettingsToggle` that should handle most of the design for you already 

Screenshot for reference to see the differences: 
![Screenshot 2024-07-31 at 5 30 56 PM](https://github.com/user-attachments/assets/90a8b781-8e0e-425d-8539-cf8bbe11bfda)

Souptik2001 committed:
refactor: Refactor single-use-links admin interface

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
@CarinaWolli Here you go -

(Just one question - Currently when you first toggle the Single Links open, there will be no links present and it will just display the "Create Link" button. Do we want to automatically create one link when nteh toggle is switched, or just the "Create Link" button is fine?)

<img width="1512" alt="Screenshot 2024-07-31 at 11 45 55 PM" src="https://github.com/user-attachments/assets/2ad11874-5905-47ca-b52e-a7d3ff58b224">

Amit91848 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -418,6 +429,103 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
+            name="singleUseLinks"
+            render={({ field: { value, onChange } }) => {
+              const addSingleUseLink = () => {
+                let seed = String(formMethods.getValues("users")[0]?.id ?? team?.id);
+                // Append the last generated link to the seed for more randomized link generation.
+                if (value.length > 0) {
+                  seed = `${seed}:${value[value.length - 1]}`;
+                } else {
+                  // If this is the first one then just append a constant string to the seed.
+                  seed = `${seed}:single_use`;
+                }
```
Is all this really necessary?

Souptik2001 replied:
Actually I wrote it to give it to make the seed more unique, though it was appended/prepended with the current time.

Thinking about it now.. I did over-engineer here.. 😛 Removed it!

Amit91848 commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -1672,15 +1672,34 @@ async function handler(
-    if (hasHashedBookingLink) {
-      await prisma.hashedLink.update({
+    if (hasHashedBookingLink && reqBody.hashedLink) {
+      // Get the hashed link to check if this is to be deleted or updated.
```
NIT: You can remove these comments. Things are understood by reading the code.

Amit91848 commented on packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:
```diff
@@ -302,40 +303,84 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
-    if (!connectedLink) {
-      // create a hashed link
-      await ctx.prisma.hashedLink.upsert({
+  const connectedPrivateLink = connectedLinks.find((link) => {
```
Updating event type removes all single use links even if it wasn't changed


https://github.com/user-attachments/assets/1a0d5636-594b-4d5f-a8a7-ab30ce7ea1ee

Souptik2001 replied:
Fixed, now! Missed this testing from my side. 🙇

Amit91848 reviewed (changes requested):
Great work so far @Souptik2001 ! Left a few comments

> (Just one question - Currently when you first toggle the Single Links open, there will be no links present and it will just display the "Create Link" button. Do we want to automatically create one link when nteh toggle is switched, or just the "Create Link" button is fine?)

Automatically creating one link would be great.

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 committed:
fix: Fix settings toggle bugs

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
chore: Remove redundant inline comments

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -418,6 +429,103 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
+                            color="destructive"
+                            className="border-none"
+                            onClick={() => removeSingleUseLink(parseInt(key))}
+                          />
```
Fixed, fixed! =

This happened after adding the settings toggle. Missed this testing from my part. 🙇

Souptik2001 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -418,6 +429,103 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
+                onChange(value);
+              };
+              return (
+                <ul ref={animateRef}>
```
Fixed!

Souptik2001 commented:
Thanks @Amit91848!

Addressed the issues.

Here is a video so far, trying to cover all edge cases -

https://www.loom.com/share/621c1d9d959c4631b5e4d838eac116ec?sid=43681bd9-4251-4e7b-b8e9-7b187c21de61

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -41,6 +42,85 @@ import {
+            {value &&
+              value.map((val: string, key: number) => {
+                const singleUseURL = `${WEBSITE_URL}/d/${val}/${formMethods.getValues("slug")}`;
+                return (
+                  <div data-testid="add-single-use-link" className="mb-4 flex items-center" key={val}>
```
Can you not make this a list item?

Souptik2001 replied:
Done!

anikdhabal commented on apps/web/playwright/event-types.e2e.ts:
```diff
@@ -367,9 +367,99 @@ testBothFutureAndLegacyRoutes.describe("Event Types tests", () => {
         await unCheckDisplayLocation(page);
       });
     });
+    test("Can add single use links", async ({ page }) => {
+      await gotoFirstEventType(page);
+
+      // Go to the advanced tab.
+      await page.getByTestId("vertical-tab-event_advanced_tab_title").click();
+
+      // Check if single-use link doesn't exist initially.
+      await expect(page.locator("[data-testid=remove-single-use-link-0]")).toHaveCount(0);
+
+      await addSingleUseLink(page);
+
+      // Save the event.
+      await saveEventType(page);
+
+      // Check if the single-use link exists.
+      await expect(page.locator("[data-testid=remove-single-use-link-0]")).toBeVisible();
+    });
+    test("Can remove single use links", async ({ page }) => {
+      await gotoFirstEventType(page);
+
+      // Go to the advanced tab.
+      await page.getByTestId("vertical-tab-event_advanced_tab_title").click();
+
+      // Check if single-use link doesn't exist initially.
```
Have you test this locally?

Souptik2001 replied:
@anikdhabal Actually yes, I was about to ask this question only - Should I refactor this to match the new toggle layout, or will we remove this test.

As I was not confirmed that if we will keep this or not, therefore I didn't modify it. But yes this will fail currently, I am aware.

anikdhabal replied:
Please refactor it and add the test after completing the all task. Here you need to add test for both frontend and backend.

Souptik2001 replied:
@anikdhabal Got it. Just to reiterate to check if I didn't misunderstood anything -

- Only when the task is completed, then only I will refactor this test.
- I have already added the part, where the it is making a booking on the front-end and after that checking if the single-use links is no more present on the backend. So, is there anything else that needs to be added. This covers both front-end and backend right?

anikdhabal replied:
Pls refactor the whole test for now. Pls add test after completing the functionality and when all things are up. I will guide you that time

Souptik2001 replied:
@anikdhabal Sorry for asking again, but did you mean "remove" or "refactor" in this line -

> Pls refactor the whole test for now

anikdhabal replied:
I mean remove your changes

Souptik2001 replied:
Thanks @anikdhabal Done!

anikdhabal commented on apps/web/modules/event-types/views/event-types-single-view.tsx:
```diff
@@ -546,6 +553,8 @@ const EventTypePage = (props: EventTypeSetupProps & { allActiveWorkflows?: Workf
   const handleSubmit = async (values: FormValues) => {
     const { children } = values;
     const dirtyValues = getDirtyFields(values);
+    console.log("HI-------");
+    console.log(dirtyValues);
```
remove this

Souptik2001 replied:
Ohooo.. my bad.. Super embarrassing.. 🫣😅

anikdhabal commented:
Fix the type error

Souptik2001 committed:
fix: Remove debug statements

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Thanks @anikdhabal . Addressed the reviews and added on query.

Souptik2001 committed:
chore: Remove single-use links e2e tests

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal commented on packages/prisma/schema.prisma:
```diff
@@ -775,10 +775,11 @@ model ApiKey {
+  id           Int       @id @default(autoincrement())
+  link         String    @unique()
+  eventType    EventType @relation(fields: [eventTypeId], references: [id], onDelete: Cascade)
+  eventTypeId  Int
+  destroyOnUse Boolean   @default(false)
```
The feature defines a single-use link. Additionally, the user does not have the ability to set the `destroyOnUse` value, so the `destroyOnUse` field is unnecessary and could always be set to true for this feature

anikdhabal reviewed (changes requested):
![Screenshot 2024-08-09 163547](https://github.com/user-attachments/assets/e9e75a32-ea51-4e69-8440-bde39344d60b)


I don’t understand why we are adding a new setting for generating a single-use link when we already have a setting for generating one-time-use private links. The single-use link does the same thing as the private link, with the only difference being that it generates multiple one-time-use links. The fields and logic used for the code implementation are also similar.

Instead of creating a new setting field, can we update the current private link feature to support generating multiple single-use links? This would mean updating the code implementation to handle multiple links instead of just one.

Currently, I have two implementation approaches in mind:

1. Similar to how we handle the current implementation: clicking the button would generate a new link item. We would need to ensure that when the setting is enabled and one link is present, the system does not display the 'add new link' button after destroying that link. Instead, it would be better if at least one link item is always present, and a new hash link is generated for that item when used.

2. Alternatively, there could be one link item or container with an external regenerate button. Clicking the regenerate button would generate a new link inside the container, allowing the user to copy each link. This would work similarly to how CAPTCHA codes and regeneration functions operate.

Souptik2001 commented:
@anikdhabal I think you missed one part here -

> he single-use link does the same thing as the private link, with the only difference being that it generates multiple one-time-use links.

The current "Private Link" thing is basically a self-updating link. I.e if you use that link, a new link will be auto-generated and replace the old link.
But the new "Single-use-link" is a auto-destroy link. I.e when you use that link, that entry will be deleted.

Although I agree that if you see behind the scene both of them are doing almost the same thing (with the "private link" having one extra feature - create new link).

Therefore we need a clear segregation between the two links. Because "private link" will be only one - which will be auto-generate after every use. But "single-use" links will be multiple which will be destroyed after use.

That being said, I think I get your point, what you are trying to say combining with this [comment](https://github.com/calcom/cal.com/pull/15896#discussion_r1711297575) is "we will depricate the previous "Private Link" feature and basically all links will be single-use links? Is it?
If that is the case, then I have no problem, and yes that would improve the editorial experience and also remove the use of `destroyOnUse`, because then all will be single-use-links only.

Please let me know. 🙏

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Tagging @Amit91848 also for suggestion 👆

anikdhabal commented:
@Souptik2001I asked my team about this feature. Please wait until I get more information

Souptik2001 commented:
Ohh perfect thanks @anikdhabal !

anikdhabal commented:
@Souptik2001 got confirmation from the team, let's merge it with private link feature

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 committed:
feat: Remove private-link feature and override with single-use-links

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix type errors

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
@anikdhabal Changes done!

Private-links is now deprecated. And all links are now single-use-links.

Amit91848 commented on packages/prisma/migrations/20240723105132_add_single_use_links_to_event_types/migration.sql:
```diff
@@ -0,0 +1,2 @@
+-- AlterTable
```
You can remove this migration and the other one because this adds destroyOnUser column, and the other one just removes the destroyOnUse column. Instead run a clean migration and create only a single migration file.

Amit91848 commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -860,6 +860,10 @@
   "copy_private_link_to_event": "Copy private link to event",
   "private_link_description": "Generate a private URL to share without exposing your {{appName}} username",
+  "single_use_links_title": "Single Use Links",
+  "single_use_links_description": "Generate a single use Link, which will be destroyed once used",
```
```suggestion
  "single_use_links_description": "Generate private single use URLs, which will be destroyed once used",
```

Amit91848 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,87 @@ import {
+                              navigator.clipboard.writeText(singleUseURL);
+                              showToast(t("single_use_link_copied"), "success");
+                            }}>
+                            <Icon name="copy" className="h-4 w-4" />
```
```suggestion
                            <Icon name="copy" className="h-4 ml-1 w-4" />
```

Amit91848 commented on packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:
```diff
@@ -146,17 +141,6 @@ export default async function handleChildrenEventTypes({
   // Calculate if there are new workflows for which assigned members will get too
   const currentWorkflowIds = eventType.workflows?.map((wf) => wf.workflowId);
 
-  // Define hashedLink query input
-  const hashedLinkQuery = (userId: number) => {
-    return hashedLink
-      ? !connectedLink
-        ? { create: { link: generateHashedLink(userId) } }
-        : undefined
-      : connectedLink
-      ? { delete: true }
-      : undefined;
-  };
-
```
Why is this removed?

Souptik2001 replied:
@Amit91848 This was responsible for updating the hashedLink(s) for child event types for "managed events".

But but, we are not showing/allowing the user to edit single-use links from the parent event type, i.e the parent managed event.

So, do you think this logic to update the single-use-links of the child events is needed?

Please let me know if I am missing something. 🙏

Amit91848 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -122,7 +200,7 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
   };
 
-  const { isChildrenManagedEventType, isManagedEventType, shouldLockDisableProps } = useLockedFieldsManager({
+  const { isChildrenManagedEventType, shouldLockDisableProps } = useLockedFieldsManager({
```
Why was managedEventType handling removed?

Amit91848 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -373,72 +451,41 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
-            <Icon name="info" className="ml-1.5 h-4 w-4 cursor-pointer" />
-          </a>
-        }
-        {...shouldLockDisableProps("hashedLink")}
```
Same here ?

Souptik2001 replied:
Added it back to the singleUseLinks. Missed it. 🙇

Amit91848 reviewed (changes requested):
Hey @Souptik2001 ! Great work, however a bunch of type checks and units are breaking. Also you have removed a bunch of stuff concerning managed event types, any particular reason? It is important and we want to keep it

Amit91848 commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,87 @@ import {
+          <ul ref={animateRef}>
+            {value &&
+              value.map((val: string, key: number) => {
+                const singleUseURL = `${WEBSITE_URL}/d/${val}/${formMethods.getValues("slug")}`;
```
These links does not work for org domains, it throws 404 page does not exist. I know the implementation is same as that of private links, but I found that it is bug there too. 

https://github.com/calcom/cal.com/blob/aab9193e7bce6f87156340f5c8c151987ffdc54a/apps/web/components/eventtype/EventTypeSingleLayout.tsx#L313-L315

You should generate the URL the same way we are doing it for Copy link to event, at the top besides the save button

Souptik2001 replied:
@Amit91848 How can we get the org URLs to run on local? 😅 

For me any org URLs are not working on local.
But yeah anyhow I have fixed the domain part which was prominently visible, but not sure any other fix is needed or not. And also couldn't test that's why..

Amit91848 replied:
You can follow the steps in the readme here https://github.com/calcom/cal.com/tree/main/packages/features/ee/organizations.

Souptik2001 replied:
Thanks Amit!

Souptik2001 committed:
refactor: Remove redundant migrations

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Add managedEvents restriction

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix single-use-links for org

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Amit91848 commented:
Is this ready for review @Souptik2001 , type checks and unit tests are failing

anikdhabal reviewed (commented):
Currently, we can see available dates and select a date and time in the case of a team event type with no host assignment through a single-use link. Generally, it displays a message indicating no availability.


https://github.com/user-attachments/assets/5c713567-7d2b-46cc-b07c-d460767a3c0a

anikdhabal reviewed (commented):
![Screenshot 2024-08-19 222820](https://github.com/user-attachments/assets/fb1983c5-81e7-4c5a-b2a1-4d217ae656ea)

Need space

anikdhabal reviewed (commented):
![Screenshot 2024-08-20 003211](https://github.com/user-attachments/assets/b931707f-74b8-4e49-b892-5416e29e3465)

I think this sounds good:- `Generate private URLs without exposing the username, which will be destroyed once used`

anikdhabal reviewed (commented):
![Screenshot 2024-08-19 224215](https://github.com/user-attachments/assets/f6b382db-d598-4ca7-be99-37a29ea28916)

Currently, if I have multiple single-use links, I can only copy one link from the button. Would it be possible to implement a round-robin algorithm, where a different link is copied each time?

anikdhabal reviewed (commented):
![Screenshot 2024-08-19 222944](https://github.com/user-attachments/assets/b2899f41-454a-49c4-98b8-acac47cfda0d)

It's not looks good when this setting is enabled and no private link exists. Can we ensure there's always at least one link present? This means users couldn't delete the only link if the setting is on, and if the last link is deleted, a new one is automatically generated to replace it

anikdhabal reviewed (commented):
There's room for improvement in the managed event type. Currently, the single-use link setting can be enabled without unlocking the field for child eventtype. But no single-use links are appearing in child event types. This is not intuitive.

This field must be unlocked in child event type when the setting is enabled, and a single link should be generated in child event type

anikdhabal reviewed (changes requested):
@Souptik2001 great work so far. But I have a few suggestions functionally before looking at the code

martinbutt commented:
For meetings with multiple duration options, when using a single-use link, if the desired duration is set in the URL, e.g. https://cal.com/d/1234567890123456789012?duration=90, it is ignored and the duration is reset to the default duration for the event.

anikdhabal commented:
> For meetings with multiple duration options, when using a single-use link, if the desired duration is set in the URL, e.g. https://cal.com/d/1234567890123456789012?duration=90, it is ignored and the duration is reset to the default duration for the event.

Yeah good observation, it needs to be fix

Souptik2001 committed:
fix: Fix recommended UI changes

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix duration not respected for private links

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix team event page title

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
feat: Add logic to copy each single-use-links from event listing page

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
feat: Add single-use links support for managed event types

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 commented:
@anikdhabal Resolved all the issues above! 

A few important updates -
- The duration bug as pointed out by @martinbutt, (which was present in the existing private-links feature also), is now fixed!
- Improved single-use links behavior on managed-links! I have tested almost all the edge-case conditions, if you still get some bug please let me know.
- Implement the round-robin algorithm for selecting the single-use link from the listing page.
- I didn't face the team single-link issue you mentioned in [this](https://github.com/calcom/cal.com/pull/15896#pullrequestreview-2246254522) comment. May have been fixed with some other change. Can you please once check at your end, if you are still facing it. But I did find a problem for team event that the page title was not properly displayed (as you can see in the video you provided also). So, fixed that also!

@Amit91848 I was thinking that can we do the type check and unit test fixes at the last, once all other things are finalized? Please let me know what do you think!
> Is this ready for review @Souptik2001 , type checks and unit tests are failing

Souptik2001 committed:
fix: Fix some type errors

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
fix: Fix type errors

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal reviewed (commented):
The single-use link is locked for the child event, but I can still generate a new link and use that field

https://github.com/user-attachments/assets/d0b63000-232f-4e99-a684-0ee0cd31ea89

The proposed solution would be:-

This field must be unlocked for the child event type when the single-use link feature is enabled in the parent event.

or

If the single-use link is enabled in the parent event and the field is locked for the child event, then there should be only one auto-generated link in the child event, and users should not be able to add a new link.

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,92 @@ import {
+                            size="sm"
+                            type="button"
+                            className="hover:stroke-3 hover:text-emphasis min-w-fit !py-0 px-0 hover:bg-transparent"
+                            aria-label="copy link"
+                            onClick={() => {
+                              navigator.clipboard.writeText(singleUseURL);
+                              showToast(t("single_use_link_copied"), "success");
+                            }}>
+                            <Icon name="copy" className="ml-1 h-4 w-4" />
+                          </Button>
+                        </Tooltip>
+                      }
+                    />
+                    {value && value.length > 1 && (
+                      <Button
+                        data-testid={`remove-single-use-link-${key}`}
+                        variant="icon"
+                        StartIcon="trash-2"
+                        color="destructive"
+                        className="ml-2 border-none"
+                        onClick={() => removeSingleUseLink(key)}
+                      />
+                    )}
+                  </li>
+                );
+              })}
+            <Button
+              color="minimal"
+              StartIcon="plus"
+              onClick={addSingleUseLink}
+              data-testid="add-single-use-link-button">
+              {t("add_a_single_use_link")}
+            </Button>
+          </ul>
+        );
+      }}
+    />
+  );
+};
+
```
Abstract this out of this file and make a new file in the root of this folder, just like we have requireconfirmationcontroller.

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,92 @@ import {
 import RequiresConfirmationController from "./RequiresConfirmationController";
 import { DisableAllEmailsSetting } from "./settings/DisableAllEmailsSetting";
 
+const SingleUseLinksManager = ({
```
```suggestion

const SingleUseLinksController = ({
```

Souptik2001 committed:
refactor: Disable adding new single-use links when field is locked

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,92 @@ import {
+
+        return (
+          <ul ref={animateRef}>
+            {value &&
+              value.map((val: string, key: number) => {
```
Unnecessary typings; not really needed

Souptik2001 commented:
@anikdhabal Related to this - https://github.com/calcom/cal.com/pull/15896#pullrequestreview-2250216010

Ohh I thought that the enable disable will be lock but new links can be added even if it is locked.
But yes makes sense to disable adding new links. Disabled that. 👍 

<img width="1055" alt="Screenshot 2024-08-21 at 2 45 32 PM" src="https://github.com/user-attachments/assets/d4388fad-cdfe-4a5b-9a82-28b755f9df98">

anikdhabal commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -46,6 +47,94 @@ import {
+                const singleUseURL = `${bookerUrl}/d/${val}/${formMethods.getValues("slug")}`;
+                return (
+                  <li data-testid="add-single-use-link" className="mb-4 flex items-center" key={val}>
+                    <TextField
+                      data-testid={`single-use-link-${key}`}
```
We don't need to add unique testid for each link

anikdhabal commented:
> <img alt="Screenshot 2024-08-21 at 2 45 32 PM" width="1055" src="https://private-user-images.githubusercontent.com/39427067/359833640-d4388fad-cdfe-4a5b-9a82-28b755f9df98.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMzI3MDcsIm5iZiI6MTcyNDIzMjQwNywicGF0aCI6Ii8zOTQyNzA2Ny8zNTk4MzM2NDAtZDQzODhmYWQtY2RmZS00YTViLTlhODItMjhiNzU1ZjlkZjk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIxVDA5MjY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3ZjVmOWJlOGIyOTNlNjA3ZWRlYTljMjJlNDQ1ZmRlMmIwYjhiZTE1MTU5MjIxZDEwMjM5MTEyNDM0ZjU1MmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jqdvnqAJR1veEW0-GdFFjPTPLKS5nhNtkmI8egrlwHQ">

Hey, instead of disabling the button, can we hide it? What do you think?

Souptik2001 committed:
fix: Address feedbacks

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 committed:
refactor: Hide add new link button for locked single-use link field

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal commented on apps/web/modules/event-types/views/event-types-listing-view.tsx:
```diff
@@ -475,8 +476,11 @@ export const EventTypeList = ({
           const calLink = `${bookerUrl}/${embedLink}`;
-          const isPrivateURLEnabled = type.hashedLink?.link;
-          const placeholderHashedLink = `${WEBSITE_URL}/d/${type.hashedLink?.link}/${type.slug}`;
+          const isPrivateURLEnabled =
+            type.hashedLink && type.hashedLink.length > 0
+              ? type.hashedLink[privateLinkCopyIndices[type.slug] ?? 0]?.link
+              : false;
+          const placeholderHashedLink = `${WEBSITE_URL}/d/${isPrivateURLEnabled}/${type.slug}`;
```
This doesn't make sense to me. Currently the type of the isPrivateURLEnabled is string | false, but it's used inside url also. Instead can you only store the link inside `const isPrivateURLEnabled = type.hashedLink[privateLinkCopyIndices[type.slug] ?? 0]?.link;`

Souptik2001 replied:
@anikdhabal Without the inline checks, i.e `type.hashedLink && type.hashedLink.length > 0`, it might throw errors when accessing array indices out-of-bounds.

But yes your point also makes sense that `bool` doesn't make any sense. Therefore I have converted the fallback value to an empty string. Does that make sense now. If the value is empty the`placeholderHashedLink` will be constructed with an empty `isPrivateURLEnabled`, but will not be displayed as per the below condition, therefore I think that should be fine. Please let me know your opinion.

anikdhabal commented on apps/web/modules/event-types/views/event-types-listing-view.tsx:
```diff
@@ -1005,8 +1015,11 @@ export const InfiniteEventTypeList = ({
             const calLink = `${bookerUrl}/${embedLink}`;
-            const isPrivateURLEnabled = type.hashedLink?.link;
-            const placeholderHashedLink = `${WEBSITE_URL}/d/${type.hashedLink?.link}/${type.slug}`;
+            const isPrivateURLEnabled =
+              type.hashedLink && type.hashedLink.length > 0
+                ? type.hashedLink[privateLinkCopyIndices[type.slug] ?? 0]?.link
+                : false;
+            const placeholderHashedLink = `${WEBSITE_URL}/d/${isPrivateURLEnabled}/${type.slug}`;
```
same here

anikdhabal reviewed (commented):
I found few type mismatch for `singleUseLinks` , such as using undefined in some places and [ ] in others

Souptik2001 committed:
chore: Use empty string for single-use link fallback value condition

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
@anikdhabal Yes makes sense to hide the button itself! - https://github.com/calcom/cal.com/pull/15896#issuecomment-2301591160

Done!

Amit91848 commented on apps/web/modules/event-types/views/event-types-listing-view.tsx:
```diff
@@ -475,8 +476,11 @@ export const EventTypeList = ({
+            type.hashedLink && type.hashedLink.length > 0
+              ? type.hashedLink[privateLinkCopyIndices[type.slug] ?? 0]?.link
+              : "";
+          const placeholderHashedLink = `${WEBSITE_URL}/d/${isPrivateURLEnabled}/${type.slug}`;
```
This would again not work for org domains. Please change it to the way it is being done in Advanced Tab

Souptik2001 replied:
@Amit91848 Done!

Amit91848 commented on apps/web/modules/event-types/views/event-types-listing-view.tsx:
```diff
@@ -1005,8 +1015,11 @@ export const InfiniteEventTypeList = ({
+              type.hashedLink && type.hashedLink.length > 0
+                ? type.hashedLink[privateLinkCopyIndices[type.slug] ?? 0]?.link
+                : "";
+            const placeholderHashedLink = `${WEBSITE_URL}/d/${isPrivateURLEnabled}/${type.slug}`;
```
Same here

Amit91848 reviewed (dismissed):
- > Currently, if I have multiple single-use links, I can only copy one link from the button. Would it be possible to implement a round-robin algorithm, where a different link is copied each time?
 
   Hmm is this really necessary? Imho I guess we should just remove copy private event link and let user copy manually from the event type tab. With it's current implementation, wouldn't it start copying from first link every time you change page or reload the window? Lmk what you think


   https://github.com/user-attachments/assets/0678ed60-34e1-4c29-af66-1d754d56f1c0

    I am not sure if this feature is implemented yet, but it currently only copies the first link every time.

- Let's just disable this particular toggle of Single Use links for parent event of Managed Event type, it doesn't really do anything.

   https://github.com/user-attachments/assets/21c3b453-e9b6-4437-bff1-b6304af267b6
   
- > @Amit91848 I was thinking that can we do the type check and unit test fixes at the last, once all other things are finalized? Please let me know what do you think!

    It is better to keep fixing it as you make changes, because these things failing means something is definitely broken / not working properly / could be missing some edge cases. If you wait till the very end, it could mean changing code which was already reviewed once and we would have to keep reviewing everything again.

anikdhabal commented:
>   Hmm is this really necessary? Imho I guess we should just remove copy private event link and let user copy manually from the event type tab. With it's current implementation, wouldn't it start copying from first link every time you change page or reload the window? Lmk what you think

It works fine for me when I tested it. Are you on the latest commit? @Souptik2001, check on your side as well. If you find a problem, then fix the implementation.

> Let's just disable this particular toggle of Single Use links for parent event of Managed Event type, it doesn't really do anything.

It does currently; I’ve checked it once. It wasn’t working before, but @Souptik2001 fixed it. Check the child event, and you will see the link

Souptik2001 commented:
@Amit91848 

The round-robin private link selection algorithm works for me also same as @anikdhabal. I checked it properly after implementing it. 

> Let's just disable this particular toggle of Single Use links for parent event of Managed Event type, it doesn't really do anything.

As @anikdhabal said, yes it is responsible for enabling and disabling the single-use links of the child events. Toggle that on, save and then check the child events. You will see one link auto-generated for those, but you can't generate more.
But you can't add more links as that option is disabled.
But I have one concern here related to this last line - `But you can't add more links as that option is disabled` - Once that link is used, there will be no option for the child event owner to add more single-use links. So, I think we should enable the option to add more links. Locked means the user can't enable or disable that single-use links feature. Please let me know your thought on this @anikdhabal @Amit91848 .


@anikdhabal @Amit91848 related to the type check fixes - when I run `yarn type-check`, I get errors from all over the project, which doesn't seem related to my changes and because of that I am not able to get the errors for my changes. Am I doing something wrong here?

anikdhabal commented:
> But I have one concern here related to this last line - `But you can't add more links as that option is disabled` - Once that link is used, there will be no option for the child event owner to add more single-use links. So, I think we should enable the option to add more links. Locked means the user can't enable or disable that single-use links feature. P

That's why I gave you two possible options to choose from. I don't think I can properly explain the first option. I think it makes sense to go with the second option, the field must be unlocked for the child event type if the setting is enabled in the parent event type.

Souptik2001 commented:
@anikdhabal

> I think it makes sense to go with the second option, the field must be unlocked for the child event type if the setting is enabled in the parent event type.

Can you please elaborate a little what do you mean by "unlock for the child event type".

Previously I had unlocked the child event type, i.e user of the child event was able to add new links and remove links, but not able to toggle the enable/disable button (as the field is locked).

But as per your previous comment, I removed the "add link" button also.

So, are you suggesting something different, which I am not able to understand?

Souptik2001 committed:
fix: Fix org single-use links in listing page

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal commented:
> Can you please elaborate a little what do you mean by "unlock for the child event type".

I am telling you that it doesn't make sense to enable the setting in the parent event type without unlocking that field for the child event type. If the field is locked, it cannot be used. Therefore, if the setting is enabled, the single-use link field should be unlocked for the child event type so that users can generate links. We need to automatically enable the button when the setting is enabled, and the field should not be toggled

I am talking about this unlock field
![Screenshot 2024-08-21 211325](https://github.com/user-attachments/assets/85b6f699-b47b-4904-b03f-1fc6813a9d1a)

Pls let me know if you don't understand

Souptik2001 commented:
@anikdhabal I got your point now perfectly..

I tried it its working on front-end..


https://github.com/user-attachments/assets/9a2c7bc6-03f6-40da-b4aa-b08eac6e2cdb


 but there are many edge cases that needs to be handled like -
- Handling the opposite interaction, i.e when someone toggles the unlock button, then the disable button should toggle oppositely.
- Then when suppose someone enables the toggle (and the field is automatically unlocked), as per the current logic, if the field is unlocked the state of the parent field will be ignored (which obviously should be), so the child's toggle is not automatically enabled. And the child user has to enable the toggle. If you think this behaviour is fine, then it is ok.

Instead should we just hide the main settings toggle button for this field and just allow the parent user to unlock and lock the field?

Please let me know your thoughts on this.

Souptik2001 committed:
feat: Auto toggle unlock button for managed event's single-use links option

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

anikdhabal commented:
@Souptik2001 Yeah, that makes sense. Let's disable the toggle button for the parent event type instead of hiding it. It would be better to show an info message below or display a tooltip message when hovering hover the toggle button. Choose whichever option looks better and makes the most sense to you. User can only lock and unlock this field for child event type.

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 committed:
Revert "feat: Auto toggle unlock button for managed event's single-use links option"

This reverts commit 5a2ed6f0bf428d52de153335499f02f42ccdf18f.

Souptik2001 committed:
chore: Disable parent toggle for single-use-links

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Thanks @anikdhabal This is done!

Please let me know if the text of the Tooltip requires any change.

<img width="1512" alt="Screenshot 2024-08-25 at 11 34 39 PM" src="https://github.com/user-attachments/assets/4138776a-d6e0-4f69-8bfc-14178ed8634e">

anikdhabal commented:
> Thanks @anikdhabal This is done!
> 
> Please let me know if the text of the Tooltip requires any change.
> 
I think the long tooltip doesn’t look very good. It would be better to display an info message at the bottom of the field. Please refer to our other info messages. Again great work @Souptik2001

anikdhabal commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -861,12 +861,17 @@
   "business_days": "business days",
   "set_address_place": "Set an address or place",
   "set_link_meeting": "Set a link to the meeting",
+  "managed_event_field_parent_control_disabled": "Can't be toggled by parent, please unlock the field to control it from child",
```
How about this ? `This field cannot be toggled. It can only be unlocked for child event types to generate and use private links`

Souptik2001 replied:
Yes much better!

Souptik2001 committed:
feat: Remove tooltip and give info message below field

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Thanks @anikdhabal ! Done!

Used this kind of info message, as it was used at another place also. Please let me know if it was supposed to be something different kind of info message -

<img width="1512" alt="Screenshot 2024-08-26 at 12 34 53 AM" src="https://github.com/user-attachments/assets/b23edf02-5648-485f-94e3-0409d6b64644">

anikdhabal commented:
> Used this kind of info message, as it was used at another place also. Please let me know if it was supposed to be something different kind of info message -
> 
First the message should be  inside the setting container, and the think is there is no chance correctly pass the message, So we need to go with tooltip. Tooltip message should be:- `Can't be toggled. It can only be unlocked for child event types`

anikdhabal committed:
add tooltip

anikdhabal committed:
tooltip message

Souptik2001 commented:
@anikdhabal Ohh ok ok.. I thought we are discarding the tooltip and instead using the message. But ok, I see you have pushed the changes to bring back the tooltip..

Please let me know if anything else needs to be done. 👍

Souptik2001 committed:
fix: Fix typo

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

CarinaWolli commented:
there are type errors in `handleChildrenEventTypes.test.ts`, @Souptik2001 could you take a look please 🙏

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Souptik2001 committed:
fix: Fix typescript type errors

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Type checks fixed. Unit tests failing 😢 Checking those..

Souptik2001 committed:
fix: Fix few unit tests

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
@anikdhabal @Amit91848 Can someone please help to debug why is this failing - 

https://github.com/Souptik2001/cal.com/blob/feature-15776-single-use-link/apps/web/test/lib/handleChildrenEventTypes.test.ts#L433

So, the `eventType.update` is not called at all 😕 -

https://github.com/Souptik2001/cal.com/blob/feature-15776-single-use-link/packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts#L261

Not able to understand why it is not reaching the update section.

Amit91848 commented on packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:
```diff
@@ -244,22 +242,39 @@ export default async function handleChildrenEventTypes({
-          },
-          data: {
-            ...updatePayloadFiltered,
-            hashedLink: "hashedLink" in unlockedFieldProps ? undefined : hashedLinkQuery(userId),
-          },
-        });
-      })
-    );
+          });
+
+          const existingHashedLink = await prisma.hashedLink.findFirst({
```
My best guess is you added these extra queries and haven't mocked their values @Souptik2001

Souptik2001 replied:
@Amit91848 But when I do a console log inside this `prisma.$trasaction`, that console log is not printed only, while if I do it inside other `prisma.$transaction` they are printed.

One thing - if you see I have changed how this `prisma.$transaction` is used. I.e passed a function instead of array of `PrismaPromises` like other usecases. So, do you think this may be because of that?

Souptik2001 replied:
@Amit91848 @anikdhabal any help on this? 😁

Amit91848 replied:
> One thing - if you see I have changed how this `prisma.$transaction` is used. I.e passed a function instead of array of `PrismaPromises` like other usecases. So, do you think this may be because of that?

Worth a try, hadn't noticed this before. AFAIK transaction expects an array of queries to be executed sequentially

Amit91848 commented on packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:
```diff
@@ -244,22 +242,39 @@ export default async function handleChildrenEventTypes({
+
+          const existingHashedLink = await prisma.hashedLink.findFirst({
+            where: {
+              eventTypeId: existingEventType ? existingEventType.id : undefined,
+            },
+          });
+
+          const eventType = await prisma.eventType.update({
+            where: {
+              id: existingEventType ? existingEventType.id : undefined,
+            },
```
Also I don't follow what's the point of calling these in case exisitngEventType is undefined?

Souptik2001 replied:
@Amit91848 Yes correct! My bad.

I have added a condition now to run this if eventType is not `undefined`.

Souptik2001 committed:
refactor: Remove redundant query

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Amit91848 committed:
Merge branch 'main' of https://github.com/calcom/cal.com into pr/15896

Amit91848 committed:
merge conflict fix

Amit91848 committed:
More merge conflict fix and refactor

Amit91848 committed:
fix tests, code refactor

Amit91848 commented on packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:
```diff
@@ -146,17 +143,6 @@ export default async function handleChildrenEventTypes({
   // Calculate if there are new workflows for which assigned members will get too
   const currentWorkflowIds = eventType.workflows?.map((wf) => wf.workflowId);
 
-  // Define hashedLink query input
-  const hashedLinkQuery = (userId: number) => {
-    return hashedLink
-      ? !connectedLink
-        ? { create: { link: generateHashedLink(userId) } }
-        : undefined
-      : connectedLink
-      ? { delete: true }
-      : undefined;
-  };
```
This part is very confusing. I have completely removed it because (based on what I could understand):
- Earlier since we had only one hashedlink per eventType, in case team admin were allowing child event types to have hasedlinks, we would update child events to each have links when that settings was toggled or in case new users were added. Which means as soon as admin toggled it, we would create links and connect it for each child event.
- I don't think it makes sense to have it now because user can toggle it in child event and have the option to generate as many as they want.
- In case it was toggled, child events had links, and then the admin changed it to not allow, we would simply delete all links for each child.

**TLDR**: We don't create hashed links for child events now. Team admins only toggle the option and each user can go to their child event have the option to create them

Amit91848 commented on packages/features/ee/managed-event-types/lib/handleChildrenEventTypes.ts:
```diff
@@ -255,7 +240,12 @@ export default async function handleChildrenEventTypes({
           data: {
             ...updatePayloadFiltered,
-            hashedLink: "hashedLink" in unlockedFieldProps ? undefined : hashedLinkQuery(userId),
+            hashedLink:
+              "singleUseLinks" in unlockedFieldProps
+                ? undefined
+                : {
+                    deleteMany: {},
+                  },
```
In case admin does not allow hashed links, delete all the hashedlinks which might be present previously

Amit91848 committed:
fix: type check

Amit91848 committed:
fix: duration config and keep links when updating event type

Amit91848 committed:
fix: e2e

Alexander-Frost commented:
Hey guys -- this PR helps fix an issue we are facing as well. Can we get 👀 on it to merge this in? 

@CarinaWolli @anikdhabal

anikdhabal commented on apps/web/lib/d/[link]/[slug]/getServerSideProps.tsx:
```diff
@@ -127,6 +127,7 @@ async function getUserPageProps(context: GetServerSidePropsContext) {
         queryDuration,
         eventData.length
       ),
+      durationConfig: eventData.metadata?.multipleDuration ?? [],
```
Why are we introducing this?

Souptik2001 replied:
@anikdhabal I added this due to this bug mentioned - https://github.com/calcom/cal.com/pull/15896#issuecomment-2297672751

in this commit - https://github.com/calcom/cal.com/pull/15896/commits/0cc66d8e92c8603db7ddef8026010bcc6d14ffdb

Long PR 😅 things are becoming hazy..

anikdhabal commented on apps/web/modules/d/[link]/d-type-view.tsx:
```diff
@@ -35,6 +36,7 @@ export default function Type({
         entity={entity}
         duration={duration}
         hashedLink={hashedLink}
+        durationConfig={durationConfig}
```
I don't see any use case for it

Souptik2001 replied:
@anikdhabal Replied above. We are passing the duration config to the `Booker` component also, so that the duration param passed to the private links are also respected.

anikdhabal commented:
> Hey guys -- this PR helps fix an issue we are facing as well. Can we get 👀 on it to merge this in?
> 
 Hey @Alexander-Frost , by this week man🙏

Alexander-Frost commented:
Bumping this!

anikdhabal reviewed (changes requested):
Hey @Souptik2001 Can you replace single-use links with multiple private links in all places? Apart from this, all things are okay 🙏

Souptik2001 committed:
refactor: Replace single-use-links with multiple-private-links

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
Hey @anikdhabal, Replaced single-use-lniks with multiple-private-links (type of words) throughout the PR. ✅

anikdhabal commented:
@Souptik2001 sorry man, all things are okay now. Could you pls fix the conflicts?

Souptik2001 committed:
Merge branch 'main' into feature-15776-single-use-link

Signed-off-by: Souptik Datta <souptikdatta2001@gmail.com>

Souptik2001 commented:
@anikdhabal No problem! 🙂 

Resolved the conflicts. ✅
